synonyms.go

```diff
@@ -0,0 +1,59 @@
+package main
+
+import "errors"
+
+type Graph map[string]map[string]bool
+
+func (g Graph) ConnectNodes(node1, node2 string) {
+	directNeighbours, found := g[node1]
+	if !found {
+		g[node1] = map[string]bool{
+			node2: true,
+		}
+	} else {
+		directNeighbours[node2] = true
+	}
+
+	directNeighbours, found = g[node2]
+	if !found {
+		g[node2] = map[string]bool{
+			node1: true,
+		}
+	} else {
+		directNeighbours[node1] = true
+	}
+}
+
+func (g Graph) GetDirectNeighbours(node string) ([]string, error) {
+	neighbourNodes, found := g[node]
+	if !found {
+		return nil, errors.New("could not find node")
+	}
+
+	nodes := []string{}
+	for k := range neighbourNodes {
+		nodes = append(nodes, k)
+	}
+
+	return nodes, nil
+}
+
+func (g Graph) GetDirectAndSecondLevelNeighbours(node string) ([]string, error) {
+	directNeighbours, err := g.GetDirectNeighbours(node)
+	if err != nil {
+		return nil, err
+	}
+
+	subnodes := []string{}
+	for _, directNeighbour := range directNeighbours {
```

Comment:
That's problematic for complexity, this for-loop will increase lookup-time from O(1) for hashmap to O(N), probably still acceptable for a small number of synonyms, but should be commented on why it is acceptable.

But talking domain: Synonyms are far more often queried/read then newly added because language changes slower then people need help with homework. So I would usually prefer the trade-off:
- Faster (constant-time) lookup
- Slower insertion.

Also, this data structure produces / stores a "hierarchy" of the words - the original 1-to-1 mappings. But they are not needed, since there's not such hierarchy in synonyms to begin with. "Removing" this hierarchy is what add complexity and time here. I would move the "removal" of this hierarchy information to the insert function where the calling frequency is way lower.